### PR TITLE
Move expression tests to jest

### PR DESCRIPTION
**Pull request by birkskyum**

### CODE REVIEW

File: test/integration/expression/lib/util.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    }
+    if (s.indexOf('\u2029') >= 0) {
+        s = s.replace(/\u2029/g, '\\u2029');
+    }
+    return s;
+}
+
+const decimalSigFigs = 6;
+
+export function stripPrecision(x) {
```
Comment (HarelM) [inline test/integration/expression/lib/util.ts:19]: Is there a library we can use to do this instead of writing our own? Also is this the only place where we use this function in our tests?

File: test/integration/expression/lib/util.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    } else {
+        const stripped = {};
+        for (const key of Object.keys(x)) {
+            stripped[key] = stripPrecision(x[key]);
+        }
+        return stripped;
+    }
+}
+
+export function deepEqual(a, b) {
```
Comment (HarelM) [inline test/integration/expression/lib/util.ts:48]: I'm pretty sure I've seen this somewhere else...?

**Review by HarelM — CHANGES_REQUESTED** [review]
Very nice work! Thanks for pushing this forward!

File: test/integration/expression/expression.test.ts [hunk 3]
```diff
@@ -1,86 +1,69 @@
...
+                done(e);
+            }
 
-run('js', {ignores, tests}, (fixture) => {
+        });
+    });
+
+});
+
+function evaluateFixture(fixture) {
```
[the remark this answers is not in the document]
Comment (birkskyum) [reply test/integration/expression/expression.test.ts:66]: in that case we need to feed it: `fixture, availableImages, canonical` - but then I guess it could be pulled out.

File: test/integration/expression/lib/util.ts (hunk 2 above)
Comment (birkskyum) [reply test/integration/expression/lib/util.ts:48]: That is true, but even though the name and overall structure is the same, this is the only place where the numerical comparison says ;

```ts
        return stripPrecision(a) === stripPrecision(b);
```

I can have a look at converting all these equality functions into one at a later PR.

File: test/integration/expression/lib/geometry.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import Point from '@mapbox/point-geometry';
+
+function getPoint(coord, canonical) {
+    const p: Point = canonical.getTilePoint(MercatorCoordinate.fromLngLat({lng: coord[0], lat: coord[1]}, 0));
+    p.x = Math.round(p.x);
+    p.y = Math.round(p.y);
+    return p;
+}
+
+function convertPoint(coord, canonical, out) {
```
[the remark this answers is not in the document]
Comment (birkskyum) [reply test/integration/expression/lib/geometry.ts:12]: We can yes. I have moved to return values.

File: test/integration/expression/lib/util.ts (hunk 1 above)
Comment (birkskyum) [reply test/integration/expression/lib/util.ts:19]: We only use the function in expression tests yes. I think there might be a library we can use, but I left it because I didn't want to break anything while refactoring.

File: test/integration/expression/lib/util.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
+import compactStringify from 'json-stringify-pretty-compact';
+
+// we have to handle this edge case here because we have test fixtures for this
+// edge case, and we don't want UPDATE=1 to mess with them
+export function stringify(v) {
```
[the remark this answers is not in the document]
Comment (birkskyum) [reply test/integration/expression/lib/util.ts:5]: I think the output could be slightly different from JSON.stringify. I left it because I didn't want to break anything while refactoring.

File: test/integration/expression/lib/util.ts (hunk 2 above)
Comment (birkskyum) [reply test/integration/expression/lib/util.ts:48]: The other definitions are found in:

src/style-spec/util/deep_equal.ts
test/integration/lib/json-diff.ts

File: test/integration/expression/lib/geometry.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
+
+import MercatorCoordinate from '../../../../src/geo/mercator_coordinate';
+import Point from '@mapbox/point-geometry';
+
+function getPoint(coord, canonical) {
```
[the remark this answers is not in the document]
Comment (birkskyum) [reply test/integration/expression/lib/geometry.ts]: much better typed now. We could potentially use type guards in getGeometry

File: test/integration/expression/lib/util.ts (hunk 2 above)
Comment (birkskyum) [reply test/integration/expression/lib/util.ts]: Managed to factor it out

File: test/integration/expression/lib/util.ts (hunk 1 above)
Comment (birkskyum) [reply test/integration/expression/lib/util.ts]: It's part of our common library now, and also used in the query tests, as it is a bit better way to check numeric equality than we used previously for query tests.

**Comment by birkskyum** [comment]
@HarelM Thanks for review! I made the improvements you requested.

File: test/integration/expression/lib/geometry.ts [hunk 7]
```diff
@@ -1,61 +1,67 @@
...
+    const l: Point[] = [];
     for (let i = 0; i < line.length; i++) {
         l.push(getPoint(line[i], canonical));
     }
-    out.push(l);
+    return l;
 }
 
-function convertLines(lines, canonical, out) {
+function convertLines(lines, canonical: CanonicalTileID) {
```
Comment (HarelM) [inline test/integration/expression/lib/geometry.ts]: Missing return type

File: test/integration/expression/lib/util.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (HarelM) [reply test/integration/expression/lib/util.ts:5]: Worth opening an issue to make sure we follow up on this later. This is what we did for other comments we decided to address later and not part of the particular PR.

File: test/integration/expression/expression.test.ts [hunk 8]
```diff
@@ -191,4 +110,74 @@ run('js', {ignores, tests}, (fixture) => {
     }
 
     return result;
-});
+}
+
+function evaluateExpression (fixture, expression, compilationResult) {
```
Comment (HarelM) [inline test/integration/expression/expression.test.ts:115]: It would be great if these method also have types.

File: test/integration/expression/lib/geometry.ts (hunk 7 above)
Comment (birkskyum) [reply test/integration/expression/lib/geometry.ts]: I think they are added in the latest commit.

File: test/integration/expression/expression.test.ts (hunk 8 above)
Comment (HarelM) [reply test/integration/expression/expression.test.ts:115]: Looked at a mid commit. Not relevant anymore...

Comment (birkskyum) [reply test/integration/expression/expression.test.ts:115]: I added the structure of the anyway - it's tricky to type, but it might help a bit anyway

File: test/integration/expression/lib/util.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (birkskyum) [reply test/integration/expression/lib/util.ts:5]: Something like this? #967 967

[the remark this answers is not in the document]
Comment (HarelM) [reply test/integration/expression/lib/util.ts:5]: Yes, Great!

File: test/integration/expression/lib/geometry.ts (hunk 7 above)
Comment (HarelM) [reply test/integration/expression/lib/geometry.ts]: Not sure what happened here. In any case this is solved.

**Review by HarelM — APPROVED** [review]
Thanks for driving this!
